### ISSUE TITLE
Converting AtomicAccess to u64

### ISSUE THREAD

**Issue by blitz** [rust-vmm/vm-memory#187]
I'm trying to implement the `Bytes` trait for the abstractions of our hypervisor and I'm really struggling. My current problem is that `Bytes` has a [store](https://docs.rs/vm-memory/0.7.0/vm_memory/bytes/trait.Bytes.html#tymethod.store) method that takes the value to store as an `AtomicAccess`. Now our internal abstractions really want an `u64` to store data. That didn't seem much of an issue until I tried converting it:

```rust
    fn store<T: AtomicAccess>(
        &self,
        val: T,
        addr: MemoryRegionAddress,
        order: Ordering,
    ) -> Result<(), Self::E> {
        let actual_value: <<T as AtomicAccess>::A as AtomicInteger>::V = val.into();

        // At this point actual_value has type V that is unconstrained and thus we can't do anything with it?
        let an_u64: u64 = actual_value as u64; // Error: an `as` expression can only be used to convert between primitive types or to coerce to a specific trait object
```

So the initially innocent looking problem of getting an actual integer I can use out of `AtomicAccess` appears really complex now. Am I missing something obvious? How do others solve this?

@bonzini

**Comment by blitz** [comment]
In C++ this would be solved by specializing template functions, but this is [not possible](https://stackoverflow.com/questions/47675493/equivalent-of-specific-template-usage-in-c-for-rust) in Rust.

**Comment by blitz** [comment]
So this solution gets somewhat closer:

```rust
unsafe fn any_int_to_u64<T>(val: T) -> u64 {
    use std::mem::{size_of_val, transmute};

    match size_of_val(&val) {
        1 => transmute::<T, u8>(val) as u64,
        2 => transmute::<T, u16>(val) as u64,
        4 => transmute::<T, u16>(val) as u64,
        8 => transmute::<T, u16>(val) as u64,
        unsupported => unimplemented!("Integer size {} cannot be converted to u64", unsupported),
    }
}
```

but fails due to:

```
unsafe fn any_int_to_u64<T>(val: T) -> u64 {
    use std::mem::{size_of_val, transmute};

    match size_of_val(&val) {
        1 => transmute::<T, u8>(val) as u64,
        2 => transmute::<T, u16>(val) as u64,
        4 => transmute::<T, u16>(val) as u64,
        8 => transmute::<T, u16>(val) as u64,
        unsupported => unimplemented!("Integer size {} cannot be converted to u64", unsupported),
    }
}
```

**Comment by jiangliu** [comment]
What does this means?
```
Now our internal abstractions really want an u64 to store data.
```
When storing an AtomicU8 value to an address, the u8 value will be converted to u64 internally?

**Comment by blitz** [comment]
It seems the way from any concrete integer type to `AtomicAccess` / `AtomicInteger` is even worse, The above approach of transmuting pointer types cannot be used because `AtomicAccess` and `AtomicInteger` do not implement `Copy` and thus the pointer cannot be dereferenced.

**Comment by blitz** [comment]
Maybe a concrete problem. Consider this method from the `Bytes` trait:

```
    fn load<T: AtomicAccess>(
        &self,
        addr: MemoryRegionAddress,
        order: std::sync::atomic::Ordering,
    )
```

How would you implement a device that always returns 23?

The only way seems to somehow restrict the type of `T` (i.e. by introducing a `where T::A::V: From<u8>`), but that does not implement the trait anymore ("impl has extra requirement").

**Comment by gsserge** [comment]
Still unsafe, but might be prettier than using transmute: `AtomicAccess` has `ByteValued` as its supertrait. `ByteValued` has a method `fn as_slice(&self) -> &[u8]`, which can be used to build up a value of `u64` using `u64::from_le_bytes()`.

**Comment by blitz** [comment]
> Still unsafe, but might be prettier than using transmute: `AtomicAccess` has `ByteValued` as its supertrait. `ByteValued` has a method `fn as_slice(&self) -> &[u8]`, which can be used to build up a value of `u64` using `u64::from_le_bytes()`.

Interesting! I wonder whether the compiler can then still remove all the cruft. The "nice" thing about the transmute solution is that the ugliness is completely optimized away. Will try that out.

Is there any example code of tests that implement MMIO devices or memory chunks that interact with these interfaces? It would be interesting to see how that is solved.

**Comment by andreeaflorescu** [comment]
Some examples are available in vm-virtio:
- https://github.com/rust-vmm/vm-virtio/blob/c5ec5126a2f2c55a6d03015ac39078f35944a7be/crates/devices/virtio-blk/src/request.rs#L207
- https://github.com/rust-vmm/vm-virtio/blob/main/crates/devices/virtio-blk/src/stdio_executor.rs#L366

We're not using `Bytes` directly afaik, but I haven't checked the code in a while so I might as well be wrong :laughing:

**Comment by blitz** [comment]
@andreeaflorescu Thanks for the pointers. Indeed trying to implement `VolatileMemory` and then implementing `Bytes` on top of it has yielded much better results. So remaining `unsafe`s I think we can fix by improving our own abstractions.

Thanks for the rubber ducking here.

**Comment by blitz** [comment]
Maybe one additional thought for the guest memory abstraction: Besides converting our RAM abstraction to `&mut [u8]` nothing else I wrote to implement `GuestMemory` has been really related to our system. Yet I wrote 240 lines of code so far to interface everything. We'll eventually share our code, but until then I'm also interested in any future redesigns of this code. I'm convinced there is a simpler abstraction hiding in here.
